File: polocloud-suite/src/main/java/dev/httpmarco/polocloud/suite/utils/PortDetector.java
package dev.httpmarco.polocloud.suite.utils;

import dev.httpmarco.polocloud.api.Polocloud;
import dev.httpmarco.polocloud.api.groups.ClusterGroup;
import dev.httpmarco.polocloud.api.platform.PlatformType;
import dev.httpmarco.polocloud.suite.services.ClusterLocalService;
import dev.httpmarco.polocloud.suite.services.ClusterLocalServiceImpl;
import lombok.experimental.UtilityClass;

import java.net.InetSocketAddress;
import java.net.ServerSocket;

@UtilityClass
public class PortDetector {

    public int nextPort(ClusterGroup group) {
        var port = group.platform().type() == PlatformType.PROXY ? 25565 : 30000;
        while (isPortUsed(port)) {
            port = port + 2;
        }
        return port;
    }

    private boolean isPortUsed(int port) {
        for (final var service : Polocloud.instance().serviceProvider().findAll()) {
            if (service instanceof ClusterLocalService localService) {
                if (localService.port() == port) return true;
            }
        }
        try (final var serverSocket = new ServerSocket()) {
            serverSocket.bind(new InetSocketAddress(port));
            return false;
        } catch (Exception exception) {
            return true;
        }
    }
}

File: agent/src/main/kotlin/dev/httpmarco/polocloud/agent/utils/PortDetector.kt
package dev.httpmarco.polocloud.agent.utils

import dev.httpmarco.polocloud.agent.Agent
import dev.httpmarco.polocloud.agent.groups.AbstractGroup
import dev.httpmarco.polocloud.agent.i18n
import dev.httpmarco.polocloud.shared.properties.START_PORT
import java.net.InetSocketAddress
import java.net.ServerSocket

class PortDetector {

    companion object {
        fun nextPort(abstractGroup: AbstractGroup): Int {
            var port = abstractGroup.platform().defaultStartPort ?: 30000

            val startPortProperty = abstractGroup.properties.get(START_PORT)
            if (startPortProperty != null) {
                try {
                    port = startPortProperty
                } catch (_: NumberFormatException) {
                    i18n.error("agent.utils.port-detector.port.cast.error", startPortProperty, abstractGroup.name)
                }
            }

            while (isPortUsed(port)) {
                port += 1
            }
            return port
        }

        private fun isPortUsed(port: Int): Boolean {
            for (service in Agent.runtime.serviceStorage().findAll()) {
                if (service.port == port) {
                    return true
                }
            }
            try {
                ServerSocket().use { serverSocket ->
                    serverSocket.bind(InetSocketAddress(port))
                    return false
                }
            } catch (_: Exception) {
                return true
            }
        }
    }
}